### ISSUE TITLE
paper #894 titled "Recommending Stack Overflow Posts for Fixing Runtime Exceptions using Failure Scenario Matching"

### ISSUE THREAD

**Issue by kstolee** [researchart/fse20#109]
This issue is for discussion related to PR #19 

Author contact: @maestro-fla

**Comment by kstolee** [comment]
## Authors:

| Name | Email | GitHub ID | Corresponding author |
| :--- | :--- | :--- | :---: |
| Sonal Mahajan  | smahajan@fujitsu.com  | [maestro-fla](https://github.com/maestro-fla) | :heavy_check_mark: |
| Negarsadat Abolhassani  | abolhass@usc.edu  | [maestro-fla](https://github.com/maestro-fla) | |
| Mukul R. Prasad | mukul@fujitsu.com | [maestro-fla](https://github.com/maestro-fla) | |

**Comment by art-reviewer** [comment]
[Paper](https://github.com/maestro-fla/fse20/blob/b3a7e0e33d5ef98a314c4c18530512a83e3a5852/submissions/functional/fse-894/esecfse2020-paper894.pdf)

**Comment by anon-eval** [comment]
@maestro-fla 
Thank you for submitting the well artifacts. The paper is also a great read :) 


**TLDR**: Based on the current version of your submission, I **cannot recommend (yet)**   the Functional badge since you do not document how to use Maestro. Although I will be happy to update it, if you can mention how Maestro can be used to recommend SO posts.

The artifacts were verified on an Ubuntu 16.04 instance with 16GB RAM

I arrived at this conclusion using the following metrics
1. **Installation** ✅ : The installation runs seamlessly without any errors.
2. **Public Availability**✅ : The tool is publicly available with a DOI
3. **Documentation** 🟡  : The tool is well documented for reproducing the results but not with respect to how the tool can be run.
4. **Consistent/Complete**✅ : The results are consistent on multiple executions and all the results in the paper can be reproduced using the scripts provided.
5. **Exercisable** ❌  : Does not mention how to run Maestro. The scripts folder and corresponding documentation only holds the scripts to reproduce the results and 
6. **Evidence of verification and validation**✅ : Since part of the study is based on manual evaluation, the ratings in RQ3 has to be manually verified. That said, by following the steps mentioned in the documentation, it is possible to arrive at the appropriate SO post for a certain exception.

Thanks :)

**Comment by gousiosg** [comment]
@art-reviewer We would appreciate your review till Monday morning AoE.

**Comment by art-reviewer** [comment]
Recommendation: **No Badge**
_Available_, _Functional_ badges may be assigned with note- _"Artifacts include datasets, tool is excluded."_

## Summary
MAESTRO presents a tool to recommend relevant StackOverflow posts to developers by matching the failure scenario of Java Runtime Exceptions. The approach uses several techniques including an Abstract Program Graph representation suitable for matching semantic structure and functionality, fault localization, and code similarity matching. Users of the tool rated it highly effective.

The artifacts include:
1. **Datasets**: comprising of-
	1. 78 Java programs with bugs in Table 1.
	2. Corpus of indexed StackOverflow posts
	3. User survey data
2. **Scripts**: for-
	1. §4.4 Effectiveness Measures (Table 1, Fig. 6-7)
	2. §4.6 Configurations used to run a comaprison tool

---
## :heavy_check_mark: Available
- [x] Publicly accessible repository: [figshare](https://doi.org/10.6084/m9.figshare.11948619)
- [x] DOI: [10.6084/m9.figshare.11948619](https://doi.org/10.6084/m9.figshare.11948619)
---
## :heavy_check_mark: Functional
- [x] Consistent: The artifacts include output of the presented tool and survey data of its users.
- [ ] Complete:	The main tool presented in the paper is missing!
- [x] Documented: The datasets and scripts are adequately documented.
- [x] Exercisable: Key results were reproduced from user survey data:
	+ §4.3 Inter-rater reliability _κ_
	+ §4.4 Effectiveness Measures- Table 1, Fig. 6-7
---
### Question to Authors
* The MAESTRO tool itself is missing from the artifacts. Please explain why it could not be included?
* Do the authors intend to release the tool separately? How?

**Comment by maestro-fla** [comment]
@anon-eval and @art-reviewer: Thank you for the detailed reviews. We belong to an industrial research lab, where the plan is to release our tool (MAESTRO) commercially. Hence, unfortunately the company policy does not allow us to make the source code of our tool open-source. However, to enable other researchers to study or compare with our approach, we release our artifacts as a **data repository**, which includes raw survey data, input benchmark, indexed Stack Overflow posts, scripts for reproducing key results in the paper, etc. Thank you for your time and consideration.

**Comment by kstolee** [comment]
@anon-eval based on the author response, can you see if the scripts are in fact exercisable? That appears to be the missing item on your review checklist.

**Comment by anon-eval** [comment]
@kstolee The scripts are exercisable if the artifacts in consideration is only the data used in the paper. I would agree with @art-reviewer that available and functional badge could be awarded but with a caveat that *"Artifacts include datasets, tool is excluded."*

**Comment by gousiosg** [comment]
I applied `available`; @anon-eval and @art-reviewer please try to reach a decision on whether `functional` is also suitable.

**Comment by art-reviewer** [comment]
**Available** and **Functional** badges are both applicable for the _included_ artifacts (data only).

The concern is with the _exclusions_. Readers expect to find and use the tools presented in a paper with these badges. With the current artifacts, this may be misleading.
So, either (1) no badges should be applied for uniformity.
OR (2) both badges may be applied with an explicit **data only** note. If such a note cannot be applied, authors may be obliged to note and explain the scope of the artifacts.

**Comment by gousiosg** [comment]
@anon-eval ?

**Comment by anon-eval** [comment]
@gousiosg I'm sorry for the delay. I would prefer recommending the **available** and badge as I appreciate the data being publicly available with the results but since the tool is not present I do not think the paper warrants the **functional** badge.

**Comment by maestro-fla** [comment]
Dear chairs and reviewers, thank you for the discussion. We don't mean to contest the opinions, but out of curiosity we would like to know what constitutes for a **functional** badge? Because if availability of the tool is one of the necessary requirements, then data repository artifacts, such as those of empirical study papers, may never qualify beyond the **available** badge. On a side note, in our query to the chairs before the submission regarding our particular case, we were encouraged to submit as a data repository artifact for which a functional/reusable badge could be applicable depending on the quality/reusability of the data.

**Comment by kstolee** [comment]
The [ACM guidelines](https://www.acm.org/publications/policies/artifact-review-badging) state as a requirement for ```functional``` that: 
>  Complete: To the extent possible, all components relevant to the paper in question are included. (Proprietary artifacts need not be included. If they are required to exercise the package then this should be documented, along with instructions on how to obtain them. Proxies for proprietary data should be included so as to demonstrate the analysis.)

It sounds like the tool is not able to be made available. However, with scripts to reproduce the results based on the data, consistency with the paper and exercisable can be checked, where exercisable is described as:
> Included scripts and/or software used to generate the results in the associated paper can be successfully executed, and included data can be accessed and appropriately manipulated.

According to @art-reviewer, such scripts are available. 

According to the call for artifacts, the following is an appropriate artifact submission:
> Data repositories, which are data (e.g., logging data, system traces, survey raw data) that can be used for multiple software engineering approaches.

From this perspective, based on the reviews, it seems that the ```functional``` badge has been earned to the extent possible. ```reusable``` would not be appropriate without the tool (since reusable implies that reuse and repurposing is possible), but it seems to me, based on the reviews and ACM guidelines, that ```functional``` is appropriate. What do you think?

**Comment by gousiosg** [comment]
I agree with @kstolee's analysis. Applying `available` and `functional`.

**Comment by maestro-fla** [comment]
Thank you very much, @kstolee and @gousiosg! We really appreciate the insightful analysis provided by @kstolee.
